### ISSUE TITLE
this model takes a long time for execution can someone help me for optimize my code

### ISSUE THREAD

**Issue by Hayat21** [MaartenGr/BERTopic#859]
import numpy as np
import re
from bertopic import BERTopic
from sentence_transformers import SentenceTransformer
from gensim import corpora, models, similarities

from src.core.cleaner import delete_unnecessary_data, preprocess, stopwords_list

def detect_topics(df,sentence_model, language=None, topics_reducing=False, top_n_words=20):
    docs = df["text"].tolist()
    df['id'] = df["CodeArticle"]
    df['title'] = df['TitreArticle']
    df['description'] = df['Textearticle']
    ids = df["id"].tolist()
    

    if language is None:
        """
        Improved models:
            For English documents the default is now: "paraphrase-MiniLM-L6-v2"
            For Non-English or multi-lingual documents the default is now: "paraphrase-multilingual-MiniLM-L12-v2"
            Both models show not only great performance but are much faster!

        For even better performance, please use the following models:
            English: "paraphrase-mpnet-base-v2"
            Non-English or multi-lingual: "paraphrase-multilingual-mpnet-base-v2"
        """
        if topics_reducing:
            topic_model = BERTopic(
                embedding_model=sentence_model,
                nr_topics="auto",
                top_n_words=top_n_words,
                calculate_probabilities=False,
                verbose=True)
        else:
            topic_model = BERTopic(
            embedding_model=sentence_model,
            top_n_words=top_n_words,
            calculate_probabilities=False,
            verbose=True)
    else:
        if topics_reducing:
            topic_model = BERTopic(
                language=language,
                nr_topics="auto",
                top_n_words=top_n_words,
                calculate_probabilities=False,
                verbose=True)
        else:
            topic_model = BERTopic(
            language=language,
            top_n_words=top_n_words,
            calculate_probabilities=False,
            verbose=True)

    topics, _ = topic_model.fit_transform(docs)

    freq = topic_model.get_topic_info()
    representative_docs = []
    global_tages = []
    for i in range(len(freq)-1):
        representative_docs.append(topic_model.get_representative_docs(topic=i))
        tags = []
        for tag in topic_model.get_topic(i):
            tag = list(tag)
            if tag[0] is None or tag[0]=='':
                continue
            tag[0] = re.sub('[\s]+',' ',str(tag[0])).strip().lower()
            if tag[0]=='' or tag[0] in stopwords_list:
                continue
            tags.append((tag[0],tag[1]))
        tags.sort(key=lambda x:x[1],reverse=True)
        global_tages.append(tags)

    topic_docs = {topic: [] for topic in set(topics)}
    for topic, id in zip(topics, ids):
        topic_docs[topic].append(id)

    topics_df = freq.loc[1:,['Topic','Count']]
    topics_df.rename(columns={'Count':'article_numbers', 'Topic':'topic_id'}, inplace=True)
    topics_df['tags'] = global_tages
    topics_df['representative_docs'] = representative_docs

    news_names = []
    news_ids = []
    relevant_docs = []
    for tages, (_, docs_ids) in zip(global_tages, topic_docs.items()):

        doc_df = df[df.id.isin(docs_ids)]

        tokens = []
        for doc in doc_df.title.tolist():
            tokens.append(preprocess(str(doc)))

        dictionary = corpora.Dictionary(tokens)

        corpus_doc2bow_vectors = [dictionary.doc2bow(tok_doc) for tok_doc in tokens]
        tfidf_model = models.TfidfModel(corpus_doc2bow_vectors, id2word=dictionary, normalize=False)
        corpus_tfidf_vectors = tfidf_model[corpus_doc2bow_vectors]
        index_matrix = similarities.SparseMatrixSimilarity(corpus_tfidf_vectors, num_features= len(dictionary))

        tages.sort(key=lambda x:x[1],reverse=True)
        query = ""
        for tag in tages:
            query += tag[0]+" "
        tokens_q = preprocess(query[:-1])
        # print(tokens_q)
        # Create a bow vector for a new document (for exmaple: a query)
        query_bow_vector = dictionary.doc2bow(tokens_q)
        # Calculate (compute) TF-IDF vector of the query
        query_tfidf_vector = tfidf_model[query_bow_vector]

        sims = index_matrix[query_tfidf_vector]

        news_names.append(doc_df.iloc[np.argmax(sims)].title)
        news_ids.append(doc_df.iloc[np.argmax(sims)].id)

        # ------------------------------------------------------------------------
        tokens = []
        docs_ids = doc_df.id.tolist()
        for doc in doc_df.text.tolist():
            tokens.append(preprocess(str(doc)))

        dictionary = corpora.Dictionary(tokens)

        corpus_doc2bow_vectors = [dictionary.doc2bow(tok_doc) for tok_doc in tokens]
        tfidf_model = models.TfidfModel(corpus_doc2bow_vectors, id2word=dictionary, normalize=False)
        corpus_tfidf_vectors = tfidf_model[corpus_doc2bow_vectors]
        index_matrix = similarities.SparseMatrixSimilarity(corpus_tfidf_vectors, num_features= len(dictionary))

        tages.sort(key=lambda x:x[1], reverse=True)
        query = ""
        for tag in tages:
            query += tag[0]+" "
        tokens_q = preprocess(query[:-1])
        # print(tokens_q)
        # Create a bow vector for a new document (for exmaple: a query)
        query_bow_vector = dictionary.doc2bow(tokens_q)
        # Calculate (compute) TF-IDF vector of the query
        query_tfidf_vector = tfidf_model[query_bow_vector]
        sims = index_matrix[query_tfidf_vector]
        sims_id = list(enumerate(sims))
        sims_id.sort(key=lambda tup: tup[1], reverse=True)
        relevant = [docs_ids[sim[0]] for sim in sims_id]
        relevant_docs.append(relevant)

    topics_df["topic"] = news_names
    topics_df["topic_doc_id"] = news_ids
    topics_df['relevant_docs'] = relevant_docs

    return topics_df

**Comment by sademakn** [comment]
1) precalculate the embeddings. then use it for training the model. 
2) https://medium.com/rapids-ai/accelerating-topic-modeling-with-rapids-and-bert-models-be9909eeed2 , use GPU for HDBSCAN and UMAP like this :

```

from bertopic import BERTopic
from cuml.cluster import HDBSCAN
from cuml.manifold import UMAP

min_topic_size = 5

umap_model = UMAP(
    n_neighbors=15,
    n_components=5,
    min_dist=0.0,
    metric="cosine",
)

hdbscan_model = HDBSCAN(
    metric="euclidean",
    cluster_selection_method="eom",
    prediction_data=True,
    gen_min_span_tree=True,
)
topic_model = BERTopic(
    min_topic_size=min_topic_size,
    calculate_probabilities=True,
    verbose=True,
    umap_model=umap_model,
    hdbscan_model=hdbscan_model,
)
```

**Comment by Hayat21** [comment]
Hello thanks for your helping
How can i use it in dockerfile

Le lun. 28 nov. 2022 à 11:50, Sajjad Dehghani ***@***.***> a
écrit :

**Comment by Hayat21** [comment]
Hello,
but how can i install cuml in a docker container

**Comment by sademakn** [comment]
https://hub.docker.com/r/rapidsai/rapidsai/
https://stackoverflow.com/questions/67896648/importing-cuml-in-a-docker-image
